Tests/NIOTests/SocketChannelTest.swift

```diff
@@ -297,4 +297,90 @@ public class SocketChannelTest : XCTestCase {
             XCTFail("Unexpected error \(error)")
         }
     }
+
+    public func testPendingConnectNotificationOrder() throws {
+
+        class NotificationOrderHandler: ChannelDuplexHandler {
+            typealias InboundIn = Never
+            typealias OutboundIn = Never
+
+            private var connectPromise: EventLoopPromise<Void>?
+
+            public func channelInactive(ctx: ChannelHandlerContext) {
+                if let connectPromise = self.connectPromise {
+                    XCTAssertTrue(connectPromise.futureResult.isFulfilled)
+                } else {
+                    XCTFail("connect(...) not called before")
+                }
+            }
+
+            public func connect(ctx: ChannelHandlerContext, to address: SocketAddress, promise: EventLoopPromise<Void>?) {
+                XCTAssertNil(self.connectPromise)
+                self.connectPromise = promise
+                ctx.connect(to: address, promise: promise)
+            }
+
+            func handlerAdded(ctx: ChannelHandlerContext) {
+                XCTAssertNil(self.connectPromise)
+            }
+
+            func handlerRemoved(ctx: ChannelHandlerContext) {
+                if let connectPromise = self.connectPromise {
+                    XCTAssertTrue(connectPromise.futureResult.isFulfilled)
+                } else {
+                    XCTFail("connect(...) not called before")
+                }
+            }
+        }
+
+        let group = MultiThreadedEventLoopGroup(numThreads: 1)
+        defer { XCTAssertNoThrow(try group.syncShutdownGracefully()) }
+
+        let eventLoop = group.next()
+        let promise: EventLoopPromise<Void> = eventLoop.newPromise()
+
+        class ConnectPendingSocket: Socket {
+            let promise: EventLoopPromise<Void>
+            init(promise: EventLoopPromise<Void>) throws {
+                self.promise = promise
+                try super.init(protocolFamily: PF_INET, type: Posix.SOCK_STREAM)
+            }
+
+            override func connect(to address: SocketAddress) throws -> Bool {
+                self.promise.succeed(result: ())
+                return false
+            }
+        }
+
+        let channel = try SocketChannel(socket: ConnectPendingSocket(promise: promise), parent: nil, eventLoop: eventLoop as! SelectableEventLoop)
+
+        XCTAssertNoThrow(try channel.pipeline.add(handler: NotificationOrderHandler()).then {
+            channel.register()
```

Comment:
You can't do this, it'll cause the channel to be insta-closed because we'll register the channel and then immediately get EPOLLHUP.

Reply:
Hmm what you suggest to do here to test this ?

Reply:
Also it works me be preferable to make this consistent on Mac and Linux as on Mac it works

Reply:
We cannot make this consistent on Mac and Linux, really, because to do it requires enormously annoying insight into the socket state. Specifically, it would require that the kqueue selector can check for all FDs registered with it whether they're connected or not before it enters the selector, which is obviously untenable.

The other solution is to have the epoll selector refuse to actually register anything until something registers for either readable or writable. That's more tenable, but the code sucks.

Reply:
@Lukasa hmm ok... so any idea for writing a test that works on mac and linux ?

Reply:
I think the fix is correct and the test shows it on Mac but would love to have a test that works on both ;)